### PR TITLE
Bschnurr/unittest discovery

### PR DESCRIPTION
**Pull request by bschnurr**
Adding unittest discovery from python adaptor
Fixing unittest execution

### CODE REVIEW

File: Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/__main__.py [hunk 1]
```diff
@@ -5,19 +5,38 @@
...
 import argparse
 import sys
+#import ptvsd
 
-from . import pytest, report
+from . import pytest, report, unittest
 from .errors import UnsupportedToolError, UnsupportedCommandError
 
 
+# 5678 is the default attach port in the VS Code debug configurations
```
Comment (bschnurr) [inline Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/__main__.py]: todo: i'll remove this

File: Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/__main__.py [hunk 2]
```diff
@@ -5,19 +5,38 @@
...
+#print("Waiting for debugger attach")
+#ptvsd.enable_attach(address=('localhost', 5678), redirect_output=True)
+#ptvsd.wait_for_attach()
+#breakpoint()
+
```
Comment (bschnurr) [inline Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/__main__.py]: todo: i'll remove this commented out ptvsd code

File: Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/report.py [hunk 3]
```diff
@@ -72,3 +72,44 @@ def report_discovered(tests, parents, pretty=False, simple=False,
...
+            }
+        for cls in suite._tests:
+            try:
+                for test in cls._tests:
+                    testdata = {
+                        'id': test.id(),
+                        'name': test._testMethodName,
+                        'lineno' : test.lineno,
+                        'source' : test.source
+                        }
```
Comment (bschnurr) [inline Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/report.py]: currently these are the attributes i'm pulling from unittest discovery.

test.id() and _testMethodName came from unittests's testloader.. i had to add lineno and source in unittest/_discover.py below

File: Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/unittest/_discovery.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    
+    root={}
+
+    for suite in suites._tests:
+        for cls in suite._tests:
+            try:
+                for test in cls._tests:
+                    #Find source, lineno from TestId and add them to test object
+                    parts = test.id().split('.')
+                    error_case, error_message = None, None
```
Comment (bschnurr) [inline Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/unittest/_discovery.py]: example id      finename.classname.testfunction  where filename has no extension

File: Python/Product/TestAdapter.Executor/UnitTest/TestDiscovererUnitTest.cs [hunk 5]
```diff
@@ -75,9 +95,9 @@ internal class TestDiscovererUnitTest : IPythonTestDiscoverer {
             arguments.Add("unittest");
             arguments.Add("--");
 
-            foreach (var s in sources) {
-                arguments.Add(s);
-            }
+            arguments.Add(_settings.ProjectHome);
+            arguments.Add("test*.py");
+
```
Comment (bschnurr) [inline Python/Product/TestAdapter.Executor/UnitTest/TestDiscovererUnitTest.cs]: I need to still hook up this pattern to project settings

File: Python/Product/TestAdapter.Executor/UnitTest/TestExecutorUnitTest.cs [hunk 6]
```diff
@@ -83,64 +85,32 @@ class TestExecutorUnitTest : ITestExecutor {
...
-                if (!String.IsNullOrWhiteSpace(djangoSettings)) {
-                    projSettings.Environment["DJANGO_SETTINGS_MODULE"] = djangoSettings;
-                }
-
-                foreach (XPathNavigator searchPath in project.Select("SearchPaths/Search")) {
-                    projSettings.SearchPath.Add(searchPath.GetAttribute("value", ""));
+                try {
+                    var discovery = DiscovererFactory.GetDiscoverer(settings);
+                    discovery.DiscoverTests(testGroup, frameworkHandle, testColletion);
+                } catch (Exception ex) {
```
Comment (bschnurr) [inline Python/Product/TestAdapter.Executor/UnitTest/TestExecutorUnitTest.cs]: when vstest execution is called from commandline we only get source files and we need to do discovery first and store tests in a testCollection.

File: Python/Product/TestAdapter.Executor/UnitTest/TestExecutorUnitTest.cs [hunk 7]
```diff
@@ -805,87 +779,7 @@ out testMethod
...
-                Tests = new List<TestCase>();
-            }
-
-            public void SendTestCase(TestCase discoveredTest) {
-                Tests.Add(discoveredTest);
-            }
-        }
-
-        sealed class PythonProjectSettings : IEquatable<PythonProjectSettings> {
-            public readonly string ProjectHome, WorkingDirectory, InterpreterPath, PathEnv;
```
Comment (bschnurr) [inline Python/Product/TestAdapter.Executor/UnitTest/TestExecutorUnitTest.cs]: this deleted code has already been moved to RunSettingsUtil.. just updating this file to use the new stuff

File: Python/Product/TestAdapter.Executor/UnitTest/UnitTestExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+using Microsoft.PythonTools.Infrastructure;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.PythonTools.TestAdapter.UnitTest {
+    public static class UnitTestExtensions {
+
+        public static TestCase ToVsTestCase(this UnitTestTestCase test, bool isWorkspace, string projectHome) {
+            var normalizedPath = test.Source.ToLowerInvariant();
```
Comment (bschnurr) [inline Python/Product/TestAdapter.Executor/UnitTest/UnitTestExtensions.cs]: after checking out the javascript test adapter I sort of liked how they converted discovered tests to vsTests using an extension class.. https://github.com/mmanela/chutzpah

I'm now thinking of updating the pytest discover flow to do something similar

File: Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/__main__.py (hunk 2 above)
Comment (bschnurr) [reply Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/__main__.py]: updated

File: Python/Product/TestAdapter.Executor/UnitTest/TestDiscovererUnitTest.cs [hunk 9]
```diff
@@ -22,50 +26,66 @@ internal class TestDiscovererUnitTest : IPythonTestDiscoverer {
...
+        private void CreateVsTests(List<UnitTestDiscoveryResults> unitTestResults, IMessageLogger logger, ITestCaseDiscoverySink discoverySink) {
+            foreach (var test in unitTestResults.SelectMany(result => result.Tests.Select(test => test))) {
+                try {
+                    var testcase = test.ToVsTestCase(_settings.IsWorkspace, _settings.ProjectHome);
+                    logger.SendMessage(TestMessageLevel.Informational, $"{testcase.DisplayName} Source:{testcase.Source} Line:{testcase.LineNumber}");
+
+                    if (discoverySink != null) {
+                        discoverySink.SendTestCase(testcase);
+                    }
```
Comment (huguesv) [inline Python/Product/TestAdapter.Executor/UnitTest/TestDiscovererUnitTest.cs]: You can use `?` and get rid of the `if`
`discoverySink?.SendTestCase()`

File: Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/__main__.py [hunk 10]
```diff
@@ -6,18 +6,26 @@
 import argparse
 import sys
 
-from . import pytest, report
+from . import pytest, report, unittest
```
[the remark this answers is not in the document]
Comment (bschnurr) [reply Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/__main__.py]: good catch. i'll see how I can fix this in another change

File: Python/Product/TestAdapter.Executor/UnitTest/TestDiscovererUnitTest.cs [hunk 11]
```diff
@@ -22,50 +26,66 @@ internal class TestDiscovererUnitTest : IPythonTestDiscoverer {
...
+                    DebugInfo("set " + _settings.PathEnv + "=" + env[_settings.PathEnv]);
+                    DebugInfo(proc.Arguments);
+
+                    // If there's an error in the launcher script,
+                    // it will terminate without connecting back.
+                    WaitHandle.WaitAny(new WaitHandle[] { proc.WaitHandle });
+
+                    outputStream.Flush();
+                    outputStream.Seek(0, SeekOrigin.Begin);
+                    var json = new StreamReader(outputStream).ReadToEnd();
```
[the remark this answers is not in the document]
Comment (bschnurr) [reply Python/Product/TestAdapter.Executor/UnitTest/TestDiscovererUnitTest.cs]: agreed

File: Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/report.py [hunk 12]
```diff
@@ -72,3 +72,44 @@ def report_discovered(tests, parents, pretty=False, simple=False,
...
+                    root['tests'].append(testdata)
```
[the remark this answers is not in the document]
Comment (bschnurr) [reply Python/Product/TestAdapter.Executor/PythonFiles/testing_tools/adapter/report.py]: will do

**Pull request merged**